src/main/java/com/minecolonies/core/items/ItemColonyFlagBanner.java

```diff
@@ -46,49 +48,35 @@ public InteractionResult useOn(UseOnContext context)
         BlockEntity te = context.getLevel().getBlockEntity(context.getClickedPos());
-        BlockState state = context.getLevel().getBlockState(context.getClickedPos());
-        ItemStack stack = context.getPlayer().getMainHandItem();
+        ItemStack stack = context.getItemInHand();
 
         if (te instanceof BannerBlockEntity || te instanceof TileEntityColonyFlag)
         {
-            CompoundTag source;
-            if (te instanceof BannerBlockEntity)
-            {
-                source = ((BannerBlockEntity) te).getItem()
-                           .getTag().getCompound("BlockEntityTag");
-            }
-            else
-            {
-                source = (context.getLevel().isClientSide ? ((TileEntityColonyFlag) te).getItemClient() : ((TileEntityColonyFlag) te).getItemServer())
-                           .getTag().getCompound("BlockEntityTag");
-            }
-
-            ListTag patternList = source.getList(TAG_BANNER_PATTERNS, 10);
-
-            // Set the base pattern, if there wasn't one set.
-            // This saves us attempting to alter the item itself to change the base color.
-            if (!patternList.getCompound(0).getString(TAG_SINGLE_PATTERN).equals(BannerPatterns.BASE.location().toString()))
-            {
-                CompoundTag nbt = new CompoundTag();
-                nbt.putString(TAG_SINGLE_PATTERN, BannerPatterns.BASE.location().toString());
-                nbt.putInt(TAG_PATTERN_COLOR, ((AbstractBannerBlock) state.getBlock()).getColor().getId());
-                patternList.add(0, nbt);
-            }
-
-            CompoundTag tag = stack.getOrCreateTagElement("BlockEntityTag");
-            tag.put(TAG_BANNER_PATTERNS, patternList);
-
-            return InteractionResult.SUCCESS;
+	    BannerPattern.Builder patternsBuilder = new BannerPattern.Builder();
+	    List<Pair<Holder<BannerPattern>, DyeColor>> source;
+	    
+	    if (te instanceof BannerBlockEntity)
+		    source = ((BannerBlockEntity) te).getPatterns();
+	    else
+		    source = ((TileEntityColonyFlag) te).getPatterns();
+	    
+	    CompoundTag bannerPattern  = new CompoundTag();
+	    source.forEach((pattern) -> patternsBuilder.addPattern(pattern.getFirst(), pattern.getSecond()));
+	    bannerPattern.put(TAG_BANNER_PATTERNS, patternsBuilder.toListTag());
+            
+	    stack.addTagElement("BlockEntityTag", bannerPattern);
+	    return InteractionResult.SUCCESS;
         }
         return super.useOn(context);
     }
 
     @Override
     public void appendHoverText(ItemStack stack, @Nullable Level worldIn, List<Component> tooltip, TooltipFlag flagIn)
     {
-        super.appendHoverText(stack, worldIn, tooltip, flagIn);
-
-        // Remove the base, as they have no translations (Mojang were lazy. Or maybe saving space?)
-        if (tooltip.size() > 1) tooltip.remove(1);
+	CompoundTag tag = BlockItem.getBlockEntityData(stack);
+	if (tag != null && tag.contains(TAG_BANNER_PATTERNS))
```

Comment:
Please follow the style guide (import the Minecolonies.xml into IntelliJ) or any of the other files for other editors.